### PR TITLE
Added RTR functionalities

### PR DESCRIPTION
**Pull request by spearmin10**
## Status
- [] In Progress
- [x] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: https://github.com/demisto/content/pull/6838

## Description
Added the CrowdStrike FalconHost RTR functionality.

## Minimum version of Demisto
- [ ] 4.5.0
- [x] 5.0.0
- [x] 5.5.0

## Does it break backward compatibility?
   - [ ] Yes
       - Further details:
   - [x] No

## Must have
- [x] Tests
- [x] Documentation 

## Demisto Partner?
- [ ] The title must be in the following format: **[YOUR_PARTNER_ID] short description**

### CODE REVIEW

File: Packs/CrowdStrikeFalcon/Integrations/CrowdStrikeFalcon/CrowdStrikeFalcon.py [hunk 1]
```diff
@@ -35,10 +38,10 @@
 INCIDENTS_PER_FETCH = int(demisto.params().get('incidents_per_fetch', 15))
 # Remove proxy if not set to true in params
 if not demisto.params().get('proxy'):
-    del os.environ['HTTP_PROXY']
```
Comment (Itay4) [inline Packs/CrowdStrikeFalcon/Integrations/CrowdStrikeFalcon/CrowdStrikeFalcon.py:38]: why did you change that ?

File: Packs/CrowdStrikeFalcon/Integrations/CrowdStrikeFalcon/CrowdStrikeFalcon.py [hunk 2]
```diff
@@ -317,7 +323,35 @@ def get_passed_mins(start_time, end_time_str):
...
+        :param host_id: Host agent ID to initialize a RTR session on.
+        :return: The session ID to execute the command on
+    """
+    endpoint_url = '/real-time-response/entities/sessions/v1'
+    body = json.dumps({
+        'device_id': host_id
+    })
+    response = http_request('POST', endpoint_url, data=body)
+    resources = response.get('resources')
+    return resources[0].get('session_id') if isinstance(resources,list) and len(resources) != 0 else None
```
Comment (Itay4) [inline Packs/CrowdStrikeFalcon/Integrations/CrowdStrikeFalcon/CrowdStrikeFalcon.py:338]: need also to check `resources[0]` is of type `dict`
lets not make it a 1-liner for it to be better readable

File: Packs/CrowdStrikeFalcon/Integrations/CrowdStrikeFalcon/CrowdStrikeFalcon.py [hunk 3]
```diff
@@ -317,7 +323,35 @@ def get_passed_mins(start_time, end_time_str):
...
+
+
+def delete_rtr_single_session(session_id: str, safe: bool = False):
+    """
+        Delete a session.
+        :param session_id: A session ID.
+        :param safe: If set to true will not raise an exception for http_error
```
Comment (Itay4) [inline Packs/CrowdStrikeFalcon/Integrations/CrowdStrikeFalcon/CrowdStrikeFalcon.py:345]: missing return in the docstring

File: Packs/CrowdStrikeFalcon/Integrations/CrowdStrikeFalcon/CrowdStrikeFalcon.py [hunk 4]
```diff
@@ -317,7 +323,35 @@ def get_passed_mins(start_time, end_time_str):
...
+    endpoint_url = '/real-time-response/entities/sessions/v1'
+    body = json.dumps({
+        'device_id': host_id
+    })
+    response = http_request('POST', endpoint_url, data=body)
+    resources = response.get('resources')
+    return resources[0].get('session_id') if isinstance(resources,list) and len(resources) != 0 else None
+
+
+def delete_rtr_single_session(session_id: str, safe: bool = False):
```
Comment (Itay4) [inline Packs/CrowdStrikeFalcon/Integrations/CrowdStrikeFalcon/CrowdStrikeFalcon.py:341]: this function is not used

File: Packs/CrowdStrikeFalcon/Integrations/CrowdStrikeFalcon/CrowdStrikeFalcon.py [hunk 5]
```diff
@@ -378,15 +427,194 @@ def run_batch_admin_cmd(host_ids: list, command_type: str, full_command: str):
...
     body = json.dumps({
         'base_command': command_type,
         'batch_id': batch_id,
         'command_string': full_command
     })
     response = http_request('POST', endpoint_url, data=body)
+    return response
+
+
+def run_batch_get_cmd(host_ids: list, file_path:str, optional_hosts:list = None, timeout:int = None, timeout_duration:str = None) -> Dict:
```
Comment (Itay4) [inline Packs/CrowdStrikeFalcon/Integrations/CrowdStrikeFalcon/CrowdStrikeFalcon.py:441]:
```suggestion
def run_batch_get_cmd(host_ids: list, file_path: str, optional_hosts: list = None, timeout: int = None, timeout_duration: str = None) -> Dict:
```

File: Packs/CrowdStrikeFalcon/Integrations/CrowdStrikeFalcon/CrowdStrikeFalcon.py [hunk 6]
```diff
@@ -378,15 +427,194 @@ def run_batch_admin_cmd(host_ids: list, command_type: str, full_command: str):
...
+    """
+        Batch executes `get` command across hosts to retrieve files.
+        After this call is made `/real-time-response/combined/get-command-status/v1` is used to query for the results.
+
+      :param host_ids: List of host agent ID’s to run RTR command on.
+      :param file_path: Full path to the file that is to be retrieved from each host in the batch.
+      :param optional_hosts: List of a subset of hosts we want to run the command on. If this list is supplied, only these hosts will receive the command.
+      :param timeout: Timeout for how long to wait for the request in seconds
+      :param timeout_duration: Timeout duration for for how long to wait for the request in duration syntax
```
Comment (Itay4) [inline Packs/CrowdStrikeFalcon/Integrations/CrowdStrikeFalcon/CrowdStrikeFalcon.py:450]: missing return in the docstring

File: Packs/CrowdStrikeFalcon/Integrations/CrowdStrikeFalcon/CrowdStrikeFalcon.py [hunk 7]
```diff
@@ -378,15 +427,194 @@ def run_batch_admin_cmd(host_ids: list, command_type: str, full_command: str):
...
+    batch_id = init_rtr_batch_session(host_ids)
+
+    body = {
+        'batch_id': batch_id,
+        'file_path': file_path
+    }
+    if optional_hosts:
+        body['optional_hosts'] = optional_hosts
 
+    params = {k:v for k,v in {
```
Comment (Itay4) [inline Packs/CrowdStrikeFalcon/Integrations/CrowdStrikeFalcon/CrowdStrikeFalcon.py:462]: not sure why this is needed, can just intialize a dict

File: Packs/CrowdStrikeFalcon/Integrations/CrowdStrikeFalcon/CrowdStrikeFalcon.py [hunk 8]
```diff
@@ -378,15 +427,194 @@ def run_batch_admin_cmd(host_ids: list, command_type: str, full_command: str):
...
+    return response
+
+
+def status_get_cmd(request_id: str, timeout:int = None, timeout_duration:str = None) -> Dict:
+    """
+        Retrieves the status of the specified batch get command. Will return successful files when they are finished processing.
+
+      :param request_id: ID to the request of `get` command.
+      :param timeout: Timeout for how long to wait for the request in seconds
+      :param timeout_duration: Timeout duration for for how long to wait for the request in duration syntax
```
Comment (Itay4) [inline Packs/CrowdStrikeFalcon/Integrations/CrowdStrikeFalcon/CrowdStrikeFalcon.py:477]: missing return in the docstring

File: Packs/CrowdStrikeFalcon/Integrations/CrowdStrikeFalcon/CrowdStrikeFalcon.py [hunk 9]
```diff
@@ -378,15 +427,194 @@ def run_batch_admin_cmd(host_ids: list, command_type: str, full_command: str):
...
+    params = {k:v for k,v in {
+        'timeout': timeout,
+        'timeout_duration': timeout_duration
+    }.items() if v}
+
+    response = http_request('POST', endpoint_url, data=json.dumps(body), params=params)
+    return response
+
+
+def status_get_cmd(request_id: str, timeout:int = None, timeout_duration:str = None) -> Dict:
```
Comment (Itay4) [inline Packs/CrowdStrikeFalcon/Integrations/CrowdStrikeFalcon/CrowdStrikeFalcon.py:471]:
```suggestion
def status_get_cmd(request_id: str, timeout: int = None, timeout_duration: str = None) -> Dict:
```

File: Packs/CrowdStrikeFalcon/Integrations/CrowdStrikeFalcon/CrowdStrikeFalcon.yml [hunk 10]
```diff
@@ -689,6 +587,197 @@ script:
...
+      description: The full file name.
+      type: string
+    - contextPath: File.Size
+      description: The size of the file in bytes.
+      type: number
+    - contextPath: File.SHA256
+      description: The SHA-256 hash of the file.
+      type: string
+    description: Get a list of files for the specified RTR session on a host.
+  - name: cs-falcon-refresh-session
```
Comment (Itay4) [inline Packs/CrowdStrikeFalcon/Integrations/CrowdStrikeFalcon/CrowdStrikeFalcon.yml:775]: why is this command needed?

File: Packs/CrowdStrikeFalcon/Integrations/CrowdStrikeFalcon/CrowdStrikeFalcon_test.py [hunk 11]
```diff
@@ -161,9 +161,10 @@ def test_run_command_read_scope(requests_mock, mocker):
     )
     results = run_command()
     expected_results = {
-        'CrowdStrike': {
+        'CrowdStrike' : {
```
Comment (Itay4) [inline Packs/CrowdStrikeFalcon/Integrations/CrowdStrikeFalcon/CrowdStrikeFalcon_test.py:164]: revert this change

File: Packs/CrowdStrikeFalcon/Integrations/CrowdStrikeFalcon/CrowdStrikeFalcon.py [hunk 12]
```diff
@@ -1418,6 +1695,237 @@ def run_script_command():
...
+def get_extracted_file_command():
+    args = demisto.args()
+    host_id = args.get('host_id')
+    sha256 = args.get('sha256')
+    filename = args.get('filename')
+
+    response = get_extracted_file(host_id, sha256, filename)
+
+    # save an extracted file
+    ct = response.headers.get('Content-Type','').lower()
```
Comment (Itay4) [inline Packs/CrowdStrikeFalcon/Integrations/CrowdStrikeFalcon/CrowdStrikeFalcon.py:1845]:
```suggestion
    content_type = response.headers.get('Content-Type','').lower()
```

File: Packs/CrowdStrikeFalcon/Integrations/CrowdStrikeFalcon/CrowdStrikeFalcon.py [hunk 13]
```diff
@@ -1418,6 +1695,237 @@ def run_script_command():
...
+    if ct == 'application/x-7z-compressed':
+        ctd = response.headers.get('Content-Disposition','').lower()
```
Comment (Itay4) [inline Packs/CrowdStrikeFalcon/Integrations/CrowdStrikeFalcon/CrowdStrikeFalcon.py:1847]:
```suggestion
        content_disposition = response.headers.get('Content-Disposition','').lower()
```

File: Packs/CrowdStrikeFalcon/Integrations/CrowdStrikeFalcon/CrowdStrikeFalcon.py [hunk 14]
```diff
@@ -1418,6 +1695,237 @@ def run_script_command():
...
+        if ctd:
+            filename = email.message_from_string(f'Content-Disposition: {ctd}\n\n').get_filename()
+
+        if not filename:
+            filename = ''.join([random.choice(string.ascii_lowercase + string.digits) for i in range(16)]) + '.dat'
```
Comment (Itay4) [inline Packs/CrowdStrikeFalcon/Integrations/CrowdStrikeFalcon/CrowdStrikeFalcon.py:1852]: better to set the sha256 instead of some random string

File: Packs/CrowdStrikeFalcon/Integrations/CrowdStrikeFalcon/CrowdStrikeFalcon.py [hunk 15]
```diff
@@ -1253,9 +1535,7 @@ def list_scripts_command():
     human_readable = tableToMarkdown('CrowdStrike Falcon scripts', scripts)
 
     entry_context = {
-        'CrowdStrike': {
```
Comment (Itay4) [inline Packs/CrowdStrikeFalcon/Integrations/CrowdStrikeFalcon/CrowdStrikeFalcon.py:1256]: while this change makes sense, it break bc
need to document that in the release notes

File: Packs/CrowdStrikeFalcon/Integrations/CrowdStrikeFalcon/CrowdStrikeFalcon.py (hunk 1 above)
Comment (spearmin10) [reply Packs/CrowdStrikeFalcon/Integrations/CrowdStrikeFalcon/CrowdStrikeFalcon.py:38]: I chaged it when it works without the keys

Comment (Itay4) [reply Packs/CrowdStrikeFalcon/Integrations/CrowdStrikeFalcon/CrowdStrikeFalcon.py:38]: not sure i understand..

File: Packs/CrowdStrikeFalcon/Integrations/CrowdStrikeFalcon/CrowdStrikeFalcon.py (hunk 7 above)
Comment (spearmin10) [reply Packs/CrowdStrikeFalcon/Integrations/CrowdStrikeFalcon/CrowdStrikeFalcon.py:462]: Will change to
`
params = {}
if timeoutNone:
    params['timeout'] = timeout
if timeout_durationNone:
    params['timeout_duration'] = timeout_duration
`
for better readable.

File: Packs/CrowdStrikeFalcon/Integrations/CrowdStrikeFalcon/CrowdStrikeFalcon.yml (hunk 10 above)
Comment (spearmin10) [reply Packs/CrowdStrikeFalcon/Integrations/CrowdStrikeFalcon/CrowdStrikeFalcon.yml:775]: Implemented it with customer request. I didn't ask them why they need it.

File: Packs/CrowdStrikeFalcon/Integrations/CrowdStrikeFalcon/CrowdStrikeFalcon.py (hunk 7 above)
Comment (Itay4) [reply Packs/CrowdStrikeFalcon/Integrations/CrowdStrikeFalcon/CrowdStrikeFalcon.py:462]: you can use `assign_params()` for that

File: Packs/CrowdStrikeFalcon/Integrations/CrowdStrikeFalcon/CrowdStrikeFalcon.py (hunk 1 above)
Comment (spearmin10) [reply Packs/CrowdStrikeFalcon/Integrations/CrowdStrikeFalcon/CrowdStrikeFalcon.py:38]: `del os.environ['HTTP_PROXY']` will make `KeyError` when HTTP_PROXY doesn't exists in os.environ. `os.environ('HTTP_PROXY', None)` works in that case.

Comment (Itay4) [reply Packs/CrowdStrikeFalcon/Integrations/CrowdStrikeFalcon/CrowdStrikeFalcon.py:38]: better to call `handle_proxy()`

File: Packs/CrowdStrikeFalcon/Integrations/CrowdStrikeFalcon/CHANGELOG.md [hunk 16]
```diff
@@ -1,4 +1,16 @@
 ## [Unreleased]
+  - Added the following real-time response API commands:
```
Comment (Itay4) [inline Packs/CrowdStrikeFalcon/Integrations/CrowdStrikeFalcon/CHANGELOG.md:2]: we dont use changelog anymore, need to move this to the release notes
see https://xsoar.pan.dev/docs/integrations/changelog

**Comment by Itay4** [comment]
@spearmin10 please add the new commands added to the integration doc README

File: Packs/CrowdStrikeFalcon/Integrations/CrowdStrikeFalcon/CrowdStrikeFalcon_test.py [hunk 17]
```diff
@@ -1304,3 +1296,785 @@ def test_list_files(requests_mock):
...
+        "powered_by": "empower-api",
+        "query_time": 1.630543865,
+        "trace_id": "8637f34a-7202-445a-818d-816715c5b368"
+      }
+    }
+    mocker.patch.object(
+        demisto,
+        'args',
+        return_value={
+            'host_ids': ['edfd6a04ad134c4344f8fb119a3ad88e'],
```
Comment (Itay4) [inline Packs/CrowdStrikeFalcon/Integrations/CrowdStrikeFalcon/CrowdStrikeFalcon_test.py:1331]: this mock here is wrong, `args` always return strings, not arrays

File: Packs/CrowdStrikeFalcon/Integrations/CrowdStrikeFalcon/CrowdStrikeFalcon.py [hunk 18]
```diff
@@ -1418,6 +1665,239 @@ def run_script_command():
...
+                'Size': resource.get('size'),
+                'SHA256': resource.get('sha256')
+            })
+            file_standard_context.append({
+                'Name': resource.get('name'),
+                'SHA256': resource.get('sha256'),
+                'Size': resource.get('size'),
+            })
+
+    human_readable = tableToMarkdown('CrowdStrike Falcon files', files_output)
```
Comment (Itay4) [inline Packs/CrowdStrikeFalcon/Integrations/CrowdStrikeFalcon/CrowdStrikeFalcon.py:1750]: please update title here

**Comment by Itay4** [comment]
whats the difference between cs-falcon-status-get-command and cs-falcon-status-command ? its unclear when need to use each

File: Packs/CrowdStrikeFalcon/Integrations/CrowdStrikeFalcon/CrowdStrikeFalcon.py [hunk 19]
```diff
@@ -1418,6 +1665,239 @@ def run_script_command():
...
+def get_extracted_file_command():
+    args = demisto.args()
+    host_id = args.get('host_id')
+    sha256 = args.get('sha256')
+    filename = args.get('filename')
+
+    response = get_extracted_file(host_id, sha256, filename)
+
+    # save an extracted file
+    content_type = response.headers.get('Content-Type','').lower()
```
Comment (Itay4) [inline Packs/CrowdStrikeFalcon/Integrations/CrowdStrikeFalcon/CrowdStrikeFalcon.py:1816]:
```suggestion
    content_type = response.headers.get('Content-Type', '').lower()
```

File: Packs/CrowdStrikeFalcon/Integrations/CrowdStrikeFalcon/CrowdStrikeFalcon.py [hunk 20]
```diff
@@ -1418,6 +1665,239 @@ def run_script_command():
...
+    if content_type == 'application/x-7z-compressed':
```
Comment (Itay4) [inline Packs/CrowdStrikeFalcon/Integrations/CrowdStrikeFalcon/CrowdStrikeFalcon.py:1817]: can you explain the logic here? this seems to add extension `7z`, while not sure it s needed, like `exe`

File: Packs/CrowdStrikeFalcon/Integrations/CrowdStrikeFalcon/CrowdStrikeFalcon.py [hunk 21]
```diff
@@ -1418,6 +1665,239 @@ def run_script_command():
...
+        content_disposition = response.headers.get('Content-Disposition','').lower()
+        if content_disposition:
+            filename = email.message_from_string(f'Content-Disposition: {content_disposition}\n\n').get_filename()
+
+        if not filename:
+            sha256 = sha256 or hashlib.sha256(response.content).hexdigest()
+            filename = sha256.lower() + '.dat'
```
Comment (Itay4) [inline Packs/CrowdStrikeFalcon/Integrations/CrowdStrikeFalcon/CrowdStrikeFalcon.py:1824]: why `dat` ?

**Comment by Itay4** [comment]
@spearmin10 please make sure all the data in test is not sensitive and is fake

File: Packs/CrowdStrikeFalcon/Integrations/CrowdStrikeFalcon/CrowdStrikeFalcon.py [hunk 22]
```diff
@@ -1418,6 +1665,239 @@ def run_script_command():
...
+    host_id = args.get('host_id')
+
+    response = list_host_files(host_id)
+    resources: dict = response.get('resources', [])
+
+    files_output = []
+    file_standard_context = []
+    command_output = []
+
+    for resource in resources:
```
Comment (Itay4) [inline Packs/CrowdStrikeFalcon/Integrations/CrowdStrikeFalcon/CrowdStrikeFalcon.py:1842]: need to improve output for no results
image

should be something `No results found`

File: Packs/CrowdStrikeFalcon/Integrations/CrowdStrikeFalcon/CrowdStrikeFalcon.py (hunk 18 above)
Comment (spearmin10) [reply Packs/CrowdStrikeFalcon/Integrations/CrowdStrikeFalcon/CrowdStrikeFalcon.py:1750]: What do you mean by updating title. Do you mean changing 'CrowdStrike Falcon files' to something else? Please let me know why the title is not good.

**Comment by spearmin10** [comment]
> whats the difference between cs-falcon-status-get-command and cs-falcon-status-command ? its unclear when need to use each

It's needed to get status after running `cs-falcon-run-get-command`

File: Packs/CrowdStrikeFalcon/Integrations/CrowdStrikeFalcon/CrowdStrikeFalcon.py (hunk 20 above)
Comment (spearmin10) [reply Packs/CrowdStrikeFalcon/Integrations/CrowdStrikeFalcon/CrowdStrikeFalcon.py:1817]: `/real-time-response/entities/extracted-file-contents/v1` called in `get_extracted_file()` must responses a 7z file when successfully done in its specification. That is a logic to get the 7z file in the response.

**Comment by Itay4** [comment]
> > whats the difference between cs-falcon-status-get-command and cs-falcon-status-command ? its unclear when need to use each
> 
> It's needed to get status after running `cs-falcon-run-get-command`

ok so lets improve these 2 commands descriptions to mention that

**Comment by spearmin10** [comment]
Fixed the code for the review. Please review again.

**Comment by kirbles19** [comment]
@Itay4 Done.

File: Packs/CrowdStrikeFalcon/ReleaseNotes/1_1_0.md [hunk 23]
```diff
@@ -0,0 +1,15 @@
...
+    - ***cs-falcon-run-get-command***
+    - ***cs-falcon-status-get-command***
+    - ***cs-falcon-status-command***
+    - ***cs-falcon-get-extracted-file***
+    - ***cs-falcon-list-host-files***
+    - ***cs-falcon-refresh-session***
+    
+  - Added the *target* argument to the ***cs-falcon-run-command*** command to support single and batch operations.
+  - Fixed entry context keys
+  - Fixed the ***cs-falcon-get-script*** command. A script entry returned from the command replaces the entry identifying with `ID` in `CrowdStrike.Script`.
```
Comment (Itay4) [inline Packs/CrowdStrikeFalcon/ReleaseNotes/1_1_0.md:14]: @kirbles19 i see it was dropped from the release notes, the last 2 bullets are breaking bc, so need to mention it

**Comment by kirbles19** [comment]
@Itay4 Updated.
